File: php/phpsearch/tests/SearchOptionsTest.php
<?php

declare(strict_types=1);

use phpsearch\SearchException;
use phpsearch\SearchOptions;
use phpsearch\SearchSettings;
use PHPUnit\Framework\TestCase;

require_once __DIR__ . '/../src/autoload.php';

class SearchOptionsTest extends TestCase
{
    public function test_no_args()
    {
        $search_options = new SearchOptions();
        try {
            $settings = $search_options->settings_from_args([]);
        } catch (SearchException $e) {
            $this->fail($e->getMessage());
        }
        $this->assertFalse($settings->archives_only);
        $this->assertTrue($settings->colorize);
        $this->assertFalse($settings->debug);
        $this->assertFalse($settings->first_match);
        $this->assertFalse($settings->follow_symlinks);
        $this->assertFalse($settings->include_hidden);
        $this->assertEquals(0, $settings->lines_after);
        $this->assertEquals(0, $settings->lines_before);
        $this->assertFalse($settings->print_dirs);
        $this->assertFalse($settings->print_files);
        $this->assertFalse($settings->print_lines);
        $this->assertEquals(150, $settings->max_line_length);
        $this->assertFalse($settings->multi_line_search);
        $this->assertTrue($settings->print_results);
        $this->assertFalse($settings->print_usage);
        $this->assertFalse($settings->print_version);
        $this->assertTrue($settings->recursive);
        $this->assertFalse($settings->search_archives);
        $this->assertFalse($settings->unique_lines);
        $this->assertFalse($settings->verbose);
    }

    public function test_valid_args()
    {
        $search_options = new SearchOptions();
        $args = ['-x', 'php,py', '-s', 'Search', '.'];
        try {
            $settings = $search_options->settings_from_args($args);
        } catch (SearchException $e) {
            $this->fail($e->getMessage());
        }
        $this->assertCount(2, $settings->in_extensions);
        $this->assertTrue(in_array('php', $settings->in_extensions));
        $this->assertTrue(in_array('py', $settings->in_extensions));
        $this->assertCount(1, $settings->search_patterns);
        $this->assertEquals('Search', $settings->search_patterns[0]);
        $this->assertCount(1, $settings->paths);
        $this->assertEquals('.', $settings->paths[0]);
    }

    public function test_archives_only_arg()
    {
        $search_options = new SearchOptions();
        $args = ['--archivesonly'];
        $settings = $search_options->settings_from_args($args);
        $this->assertTrue($settings->archives_only);
        $this->assertTrue($settings->search_archives);
    }

    public function test_debug_arg()
    {
        $search_options = new SearchOptions();
        $args = ['--debug'];
        $settings = $search_options->settings_from_args($args);
        $this->assertTrue($settings->debug);
        $this->assertTrue($settings->verbose);
    }

    public function test_missing_arg()
    {
        $search_options = new SearchOptions();
        $this->expectException(SearchException::class);
        $args = ['-x', 'php,py', '-s', 'Search', '.', '-D'];
        $search_options->settings_from_args($args);
    }

    public function test_invalid_arg()
    {
        $search_options = new SearchOptions();
        $this->expectException(SearchException::class);
        $args = ['-x', 'php,py', '-s', 'Search', '.', '-Q'];
        $search_options->settings_from_args($args);
    }

    public function test_settings_from_json()
    {
        $settings = new SearchSettings();
        $json = <<<"END_JSON"
{
  "path": "~/src/xsearch/",
  "in-ext": ["js","ts"],
  "out-dirpattern": "node_module",
  "out-filepattern": ["temp"],
  "searchpattern": "Searcher",
  "linesbefore": 2,
  "linesafter": 2,
  "debug": true,
  "allmatches": false,
  "followsymlinks": true,
  "includehidden": true
}
END_JSON;
        $search_options = new SearchOptions();
        $search_options->update_settings_from_json($settings, $json);
        $this->assertCount(1, $settings->paths);
        $this->assertTrue(in_array('~/src/xsearch/', $settings->paths));
        $this->assertCount(2, $settings->in_extensions);
        $this->assertTrue(in_array('js', $settings->in_extensions));
        $this->assertTrue(in_array('ts', $settings->in_extensions));
        $this->assertCount(1, $settings->out_dir_patterns);
        $this->assertEquals('node_module', $settings->out_dir_patterns[0]);
        $this->assertCount(1, $settings->out_file_patterns);
        $this->assertEquals('temp', $settings->out_file_patterns[0]);
        $this->assertCount(1, $settings->search_patterns);
        $this->assertEquals('Searcher', $settings->search_patterns[0]);
        $this->assertEquals(2, $settings->lines_before);
        $this->assertEquals(2, $settings->lines_after);
        $this->assertTrue($settings->debug);
        $this->assertTrue($settings->verbose);
        $this->assertTrue($settings->first_match);
        $this->assertTrue($settings->follow_symlinks);
        $this->assertTrue($settings->include_hidden);
    }
}

File: python/pysearch/tests/searchoptions_test.py
# -*- coding: utf-8 -*-
"""
################################################################################
#
# searchoptions_test.py
#
# class SearchOptionsTest: testing of SearchOptions class
#
################################################################################
"""
import os
from pathlib import Path
import sys
import unittest

sys.path.insert(0, os.path.abspath(os.path.dirname(__file__)[:-6]))

from pysearch import SearchException, SearchOptions, SearchSettings


class SearchOptionsTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.search_options = SearchOptions()

    def test_no_args(self):
        # test the props
        settings = self.search_options.search_settings_from_args([])
        self.assertFalse(settings.archives_only)
        self.assertFalse(settings.debug)
        self.assertFalse(settings.first_match)
        self.assertFalse(settings.include_hidden)
        self.assertEqual(settings.lines_after, 0)
        self.assertEqual(settings.lines_before, 0)
        self.assertFalse(settings.print_dirs)
        self.assertFalse(settings.print_files)
        self.assertFalse(settings.print_lines)
        self.assertEqual(settings.max_line_length, 150)
        self.assertFalse(settings.multi_line_search)
        self.assertTrue(settings.print_results)
        self.assertFalse(settings.print_usage)
        self.assertFalse(settings.print_version)
        self.assertTrue(settings.recursive)
        self.assertFalse(settings.search_archives)
        self.assertFalse(settings.unique_lines)
        self.assertFalse(settings.verbose)
        # test the extension and pattern sets
        self.assertFalse(settings.in_archive_extensions)
        self.assertFalse(settings.in_archive_file_patterns)
        self.assertFalse(settings.in_dir_patterns)
        self.assertFalse(settings.in_file_patterns)
        self.assertFalse(settings.in_lines_after_patterns)
        self.assertFalse(settings.in_lines_before_patterns)
        self.assertFalse(settings.lines_after_to_patterns)
        self.assertFalse(settings.lines_after_until_patterns)
        self.assertFalse(settings.out_archive_extensions)
        self.assertFalse(settings.out_archive_file_patterns)
        self.assertFalse(settings.out_dir_patterns)
        self.assertFalse(settings.out_file_patterns)
        self.assertFalse(settings.out_lines_after_patterns)
        self.assertFalse(settings.out_lines_before_patterns)
        self.assertFalse(settings.search_patterns)

    def test_valid_args(self):
        args = ['-x', 'py,rb', '-s', 'Search', '.']
        settings = self.search_options.search_settings_from_args(args)
        self.assertEqual(1, len(settings.paths))
        for x in {'py', 'rb'}:
            self.assertIn(x, settings.in_extensions)
        self.assertEqual('Search', list(settings.search_patterns)[0].pattern)

    def test_archives_only_arg(self):
        args = ['--archivesonly']
        settings = self.search_options.search_settings_from_args(args)
        self.assertTrue(settings.archives_only)
        self.assertTrue(settings.search_archives)

    def test_debug_arg(self):
        args = ['--debug']
        settings = self.search_options.search_settings_from_args(args)
        self.assertTrue(settings.debug)
        self.assertTrue(settings.verbose)

    def test_missing_arg(self):
        args = ['-x', 'py,rb', '-s', 'Search', '.', '-D']
        with self.assertRaises(SearchException) as cm:
            settings = self.search_options.search_settings_from_args(args)
        self.assertEqual(str(cm.exception), 'Missing value for option D')

    def test_invalid_arg(self):
        args = ['-x', 'py,rb', '-s', 'Search', '.', '-Q']
        with self.assertRaises(SearchException) as cm:
            settings = self.search_options.search_settings_from_args(args)
        self.assertEqual(str(cm.exception), 'Invalid option: Q')

    def test_settings_from_json(self):
        settings = SearchSettings()
        json = '''{
  "path": "~/src/xsearch/",
  "in-ext": ["js","ts"],
  "out-dirpattern": ["node_module", "dist"],
  "out-filepattern": ["temp"],
  "searchpattern": "Searcher",
  "linesbefore": 2,
  "linesafter": 2,
  "debug": true,
  "allmatches": false,
  "includehidden": true
}'''
        self.search_options.settings_from_json(json, settings)
        self.assertEqual(1, len(settings.paths))
        self.assertIn(Path('~/src/xsearch/'), settings.paths)
        for x in {'js', 'ts'}:
            self.assertIn(x, settings.in_extensions)
        self.assertTrue(any([p.pattern == 'Searcher' for p in settings.search_patterns]))
        self.assertTrue(any([p.pattern == 'node_module' for p in settings.out_dir_patterns]))
        self.assertTrue(any([p.pattern == 'dist' for p in settings.out_dir_patterns]))
        self.assertTrue(any([p.pattern == 'temp' for p in settings.out_file_patterns]))
        self.assertEqual(settings.lines_before, 2)
        self.assertEqual(settings.lines_after, 2)
        self.assertTrue(settings.debug)
        self.assertTrue(settings.verbose)
        self.assertTrue(settings.first_match)
        self.assertTrue(settings.include_hidden)


if __name__ == '__main__':
    unittest.main()
